src/Compilers/Core/Portable/Operations/ControlFlowGraph.Region.cs

```diff
@@ -0,0 +1,183 @@
+            /// 
+            /// An <see cref="ITryOperation"/> that has a set of <see cref="ITryOperation.Catches"/> and a <see cref="ITryOperation.Finally"/> 
+            /// at the same time is mapped to a <see cref="TryAndFinally"/> region with <see cref="TryAndCatch"/> region inside its <see cref="Try"/> region.
+            /// </summary>
+            TryAndFinally,
+        }
+
+        /// <summary>
+        /// Encapsulates information about regions of <see cref="BasicBlock"/>s in a <see cref="ControlFlowGraph"/>.
+        /// Regions can overlap, but never cross each other boundaries.
+        /// </summary>
+        public sealed class Region
+        {
+            /// <summary>
+            /// Region's kind
+            /// </summary>
+            public RegionKind Kind { get; }
+
+            /// <summary>
+            /// Enclosing region. Null for <see cref="RegionKind.Root"/>
+            /// </summary>
+            public Region Enclosing { get; private set; }
+
+            /// <summary>
+            /// Target exception type for <see cref="RegionKind.Filter"/>, <see cref="RegionKind.Handler"/>, 
+            /// <see cref="RegionKind.FilterAndHandler "/>
+            /// </summary>
+            public ITypeSymbol ExceptionType { get; }
+
+            /// <summary>
+            /// Ordinal (<see cref="BasicBlock.Ordinal"/>) of the first <see cref="BasicBlock"/> within the region. 
+            /// </summary>
+            public int FirstBlockOrdinal { get; }
+
+            /// <summary>
+            /// Ordinal (<see cref="BasicBlock.Ordinal"/>) of the last <see cref="BasicBlock"/> within the region. 
+            /// </summary>
+            public int LastBlockOrdinal { get; }
+
+            /// <summary>
+            /// Regions within this region
+            /// </summary>
+            public ImmutableArray<Region> Regions { get; }
+
+            /// <summary>
+            /// Locals for which this region represent the life-time.
+            /// </summary>
+            public ImmutableArray<ILocalSymbol> Locals { get; }
+
+            internal Region(RegionKind kind, int firstBlockOrdinal, int lastBlockOrdinal, 
+                            ImmutableArray<Region> regions = default, ImmutableArray<ILocalSymbol> locals = default, 
+                            ITypeSymbol exceptionType = null)
+            {
+                Debug.Assert(firstBlockOrdinal >= 0);
+                Debug.Assert(lastBlockOrdinal >= firstBlockOrdinal);
+
+                Kind = kind;
+                FirstBlockOrdinal = firstBlockOrdinal;
+                LastBlockOrdinal = lastBlockOrdinal;
+                ExceptionType = exceptionType;
+                Locals = locals.NullToEmpty();
+                Regions = regions.NullToEmpty();
+
+
+                foreach(Region r in Regions)
+                {
+                    Debug.Assert(r.Enclosing == null && r.Kind != RegionKind.Root);
```

Comment:
Perhaps assert that `r.Enclosing is null` (it should be set only once).

Reply:
> Perhaps assert that r.Enclosing is null (it should be set only once).

Could you please clarify? It looks like the assert verifies that r.Enclosing is null. 

---
In reply to: [167941493](https://github.com/dotnet/roslyn/pull/24770#discussion_r167941493) [](ancestors = 167941493)

Reply:
Maybe I should drink coffee after all ;-)
Got confused

---
In reply to: [167962169](https://github.com/dotnet/roslyn/pull/24770#discussion_r167962169) [](ancestors = 167962169,167941493)